tamr_unify_client/dataset/resource.py

```diff
@@ -434,3 +444,88 @@ def _geo_attr_names():
             "polygon",
             "multiPolygon",
         }
+
+
+class DatasetSpec:
+    """A representation of the server view of a dataset."""
+
+    def __init__(self, client, data, api_path):
+        self._data = data
+        self.client = client
+        self.api_path = api_path
+
+    @staticmethod
+    def of(dataset):
+        """Creates a dataset spec from a dataset.
+
+        :param dataset: The existing dataset.
+        :type dataset: :class:`~tamr_unify_client.dataset.resource.Dataset`
+        :return: The corresponding dataset spec.
+        :rtype: :class:`~tamr_unify_client.dataset.resource.DatasetSpec`
+        """
+        return DatasetSpec(dataset.client, deepcopy(dataset._data), dataset.api_path)
+
+    def from_data(self, data):
+        """Creates a spec with the same client and API path as this one, but new data.
+
+        :param data: The data for the new spec.
+        :type data: dict
+        :return: The new spec.
+        :rtype: :class:`~tamr_unify_client.dataset.resource.DatasetSpec`
+        """
+        return DatasetSpec(self.client, data, self.api_path)
+
+    def to_dict(self):
+        """Returns a copy of this spec's internal representation.
+
+        :returns: The spec's dict.
+        :rtype: dict
+        """
+        return deepcopy(self._data)
+
+    def with_external_id(self, new_external_id):
+        """Creates a new spec with the same properties, updating external ID.
+
+        :param new_external_id: The new external ID.
+        :type new_external_id: str
+        :return: A new spec.
+        :rtype: :class:`~tamr_unify_client.dataset.resource.DatasetSpec`
+        """
+        return self.from_data({**self._data, "externalId": new_external_id})
+
+    def with_description(self, new_description):
+        """Creates a new spec with the same properties, updating description.
+
+        :param new_description: The new description.
+        :type new_description: str
+        :return: A new spec.
+        :rtype: :class:`~tamr_unify_client.dataset.resource.DatasetSpec`
+        """
+        return self.from_data({**self._data, "description": new_description})
+
+    def with_tags(self, new_tags):
+        """Creates a new spec with the same properties, updating tags.
+
+        :param new_tags: The new tags.
+        :type new_tags: list[str]
```

Comment:
`list[str]` isn't our contract; it's the contract of the API. My preference would be to explicitly defer to the API contract, but the reality is that contract is supposed to not change, so it's both low risk and user friendly to replicate it here.

Reply:
What @nbateshaus and I decided:
If a property of a resource is a complex type (like an attribute type), it should have its own spec class, and a user should provide an instance of this spec to `with_*`.
If a property is not a complex type (e.g. a string), it would not have a spec class, and the user would provide the simple representation directly.
If the property is a list/set of resources, the user would provide a list of these resources, represented as determined by the above.

A tag is represented as a string, so there should not be a resource or spec class for a tag. Consequently, a user should just provide a list of strings to `with_tags`.